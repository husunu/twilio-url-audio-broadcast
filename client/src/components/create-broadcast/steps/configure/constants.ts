export const buyNewNumberUrl =
  'https://www.twilio.com/console/phone-numbers/search';
