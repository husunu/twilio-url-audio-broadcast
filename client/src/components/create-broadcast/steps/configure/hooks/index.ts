export { useConfigureState } from './useConfigureState';
export { useNumbers } from './useNumbers';
export { useTestCall } from './useTestCall';
