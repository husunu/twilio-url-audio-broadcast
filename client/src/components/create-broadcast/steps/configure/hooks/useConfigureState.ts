import { useCallback, useEffect, useMemo, useState } from 'react';
import isEmpty from 'lodash.isempty';
import { ConfigureStep } from '@/types';
import { AlertDto } from '@/components/common/alert';

export const useConfigureState = (data: ConfigureStep) => {
  const [from, setFrom] = useState(data.number);
  const [message, setMessage] = useState(data.message);
  const [name, setName] = useState(data.name);
  const [testNumber, setTestNumber] = useState<string>('');
  const [alert, setAlert] = useState<AlertDto>({});

  useEffect(() => {
    setFrom(data.number);
    setMessage(data.message);
    setName(data.name);
  }, [data]);

  const isValid = useMemo(() => {
    return !isEmpty(name) && !isEmpty(message) && !isEmpty(from);
  }, [from, message, name]);

  const resetAlert = useCallback(() => {
    setAlert({});
  }, [setAlert]);

  return {
    alert,
    setAlert,
    resetAlert,
    isValid,
    from,
    setFrom,
    message,
    setMessage,
    name,
    setName,
    testNumber,
    setTestNumber,
  };
};
