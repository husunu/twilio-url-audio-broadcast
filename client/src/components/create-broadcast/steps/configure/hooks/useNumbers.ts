import { useCallback, useEffect, useState } from 'react';
import useFetch from 'use-http';
import { IncomingPhoneNumberInstance } from 'twilio/lib/rest/api/v2010/account/incomingPhoneNumber';

export const useNumbers = () => {
  const [numbers, setNumbers] = useState<IncomingPhoneNumberInstance[]>([]);
  const { request } = useFetch<IncomingPhoneNumberInstance[]>();

  const loadNumbers = useCallback(() => {
    (async () => {
      const response = await request.get('/numbers');
      setNumbers(response);
    })();
  }, [setNumbers, request]);

  useEffect(() => {
    loadNumbers();
  }, [request, loadNumbers]);
  return { numbers, loadNumbers };
};
