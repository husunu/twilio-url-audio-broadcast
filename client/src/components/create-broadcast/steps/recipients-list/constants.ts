export const MAX_NUMBERS = 500;

export const ALERTS = {
  FILE_READ_ERROR: 'There was an error reading the file.',
  FILE_EXCEEDS_MAX: `File has more than ${MAX_NUMBERS} numbers. Please review issues and upload again.`,
  INVALID_NUMBERS:
    'The list of recipients has invalid numbers. Please review issues and upload again.',
  PARSE_SUCCESS: 'The list of recipients is successfully loaded.',
  FILE_REJECT_ERROR: 'Please make sure you tried to upload a text file (.txt)',
};

export const INVALID_NUMBERS = 'invalidNumbers';
export const VALID_NUMBERS = 'validNumbers';

export const INITIAL_VALIDATION_RESULT = {
  [INVALID_NUMBERS]: [],
  [VALID_NUMBERS]: [],
};
