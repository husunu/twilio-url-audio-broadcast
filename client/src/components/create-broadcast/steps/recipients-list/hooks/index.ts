export { useCustomDropZone } from './useCustomDropZone';
export { useFileReader } from './useFileReader';
export { useNumbersValidator } from './useNumbersValidator';
export { useAlertSetter } from './useAlertSetter';
export { useRecipientsList } from './useRecipientsList';
