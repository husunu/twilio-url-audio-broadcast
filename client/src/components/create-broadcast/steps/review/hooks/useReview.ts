import { useCallback } from 'react';
import { completeReviewStep, Steps, stepsSelector } from '@/redux';
import useFetch from 'use-http/dist';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory } from 'react-router-dom';
import { ConfigureStep, RecipientsStep } from '@/types';

type useReviewResult = {
  onComplete: () => void;
  loading: boolean;
  error?: Error;
  numbers: string[];
  number: string;
  message: string;
  name: string;
};

export const useReview = (): useReviewResult => {
  const steps: Steps = useSelector(stepsSelector);

  const { configure, recipients } = steps;
  const { message, name, number } = configure as ConfigureStep;
  const { numbers } = recipients as RecipientsStep;

  const dispatch = useDispatch();
  const history = useHistory();
  const { post, response, loading, error } = useFetch();

  const onComplete = useCallback(() => {
    (async () => {
      const { broadcastId } = await post('/broadcasts', {
        friendlyName: name,
        message,
        from: number,
        to: numbers.join(','),
      });

      if (response.ok) {
        dispatch(completeReviewStep({ broadcastId, completed: true }));
        history.push(`/broadcasts/${broadcastId}?success=true`);
      }
    })();
  }, [post, response, history, dispatch, message, name, number, numbers]);

  return { onComplete, loading, error, numbers, number, message, name };
};
