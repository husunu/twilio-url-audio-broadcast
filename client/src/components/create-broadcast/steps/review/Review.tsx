import React, { useMemo } from 'react';
import { Alert } from '@material-ui/lab';
import { Grid, Typography, CircularProgress, Box } from '@material-ui/core';
import { useReview } from './hooks/useReview';
import { TextWrap } from '@/components/common/text-wrap';
import { NumbersTable } from '@/components/common/numbers-table';
import { StyledCard } from '@/components/common/styled-card';
import ReviewNavigation from './ReviewNavigation';
import useReviewStyles from './styles';

import {
  StepProps,
  Wizard,
  VALID_NUMBERS_TABLE,
} from '@/components/create-broadcast';

export const Review = ({ step }: StepProps) => {
  const classes = useReviewStyles();
  const {
    onComplete,
    loading,
    error,
    numbers,
    number,
    message,
    name,
  } = useReview();

  const formattedNumbers = useMemo(() => {
    return numbers?.map((formattedNumber, index) => ({
      line: index + 1,
      number: formattedNumber,
    }));
  }, [numbers]);

  return (
    <>
      <Wizard.Step step={step}>
        <Typography variant="h4" component="h2">
          3. Review
        </Typography>
        <Box mt={1}>
          <Typography>
            Your voice notification is almost done. Please review it and press
            “Send Notification” to start sending to the recipient list.
          </Typography>
        </Box>
        {(loading || error) && (
          <Box className={classes.progressContainer}>
            {loading && <CircularProgress />}
            {error && (
              <Alert severity="error">
                We could not create your broadcast this time, sorry.
              </Alert>
            )}
          </Box>
        )}
        <StyledCard className={classes.cardReview}>
          <Grid container spacing={4}>
            <Grid item xs={12} md={6}>
              <Box component="dl">
                <Typography variant="subtitle2" component="h2">
                  Notification Name
                </Typography>
                <Box mb={2}>
                  <Typography variant="subtitle1">{name}</Typography>
                </Box>
                <Typography variant="subtitle2" component="h2">
                  From number
                </Typography>
                <Box mb={2}>
                  <Typography variant="subtitle1">{number}</Typography>
                </Box>
                <Typography variant="subtitle2" component="h2">
                  Notification Message
                </Typography>
                <Box mb={2}>
                  <Typography variant="body2">
                    <TextWrap text={message} />
                  </Typography>
                </Box>
              </Box>
            </Grid>
            <Grid item xs={12} md={6}>
              <Typography
                variant="subtitle2"
                component="h2"
                style={{ marginBottom: '15px' }}
              >
                Recipients
              </Typography>
              <NumbersTable
                headers={VALID_NUMBERS_TABLE.headers}
                rows={VALID_NUMBERS_TABLE.rows}
                data={formattedNumbers}
              />
            </Grid>
          </Grid>
        </StyledCard>
        <Alert severity="warning">
          The time it takes to complete sending all messages depends on the
          number of recipients and the Twilio account CPS rate (Calls Per
          Second). Please notice that by default accounts have a 1 CPS rate.
          Contact Sales or Support to increase your CPS rate.
        </Alert>
      </Wizard.Step>
      <ReviewNavigation
        onComplete={onComplete}
        canGoNext={!loading}
        currentStep={step}
      />
    </>
  );
};
