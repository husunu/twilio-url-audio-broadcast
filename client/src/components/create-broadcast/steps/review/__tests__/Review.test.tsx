import React from 'react';
import { renderWithAppContexts } from '@/testUtils';
import { Review } from '../Review';
import { steps } from '@/components/create-broadcast/constants';
import { useReview } from '../hooks/useReview';

jest.mock('../hooks/useReview');
const useReviewMock = useReview as jest.MockedFunction<typeof useReview>;

describe('Review', () => {
  afterAll(jest.clearAllMocks);
  test('render the page', () => {
    useReviewMock.mockReturnValue({
      onComplete: jest.fn(),
      loading: false,
      numbers: ['+34555666777'],
      number: '+34666222333',
      message: 'Test message',
      name: 'Test name',
    });
    const props = {
      data: {
        completed: true,
        broadcastId: '',
      },
      step: steps[2],
    };
    const wrapper = renderWithAppContexts(<Review {...props} />);
    expect(wrapper.container).toMatchSnapshot();
  });
  test('if post fails it should show an error message', () => {
    useReviewMock.mockReturnValue({
      onComplete: jest.fn(),
      loading: false,
      error: new Error('Something went wrong'),
      numbers: ['+34555666777'],
      number: '+34666222333',
      message: 'Test message',
      name: 'Test name',
    });
    const props = {
      data: {
        completed: true,
        broadcastId: '',
      },
      step: steps[2],
    };
    const { getByText } = renderWithAppContexts(<Review {...props} />);
    expect(
      getByText(/We could not create your broadcast this time, sorry./i)
    ).toBeInTheDocument();
  });
});
