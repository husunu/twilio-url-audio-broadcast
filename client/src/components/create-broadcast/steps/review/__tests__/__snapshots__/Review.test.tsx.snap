// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Review render the page 1`] = `
<div>
  <div
    class="MuiPaper-root MuiCard-root WithStyles(ForwardRef(Card))-root-5 WithStyles(ForwardRef(Card))-root-6 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="MuiBox-root MuiBox-root-37"
    >
      <h2
        class="MuiTypography-root MuiTypography-h4"
      >
        3. Review
      </h2>
      <div
        class="MuiBox-root MuiBox-root-68"
      >
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          Your voice notification is almost done. Please review it and press “Send Notification” to start sending to the recipient list.
        </p>
      </div>
      <div
        class="MuiPaper-root MuiCard-root WithStyles(ForwardRef(Card))-root-5 WithStyles(ForwardRef(Card))-root-69 makeStyles-cardReview-4 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6"
          >
            <dl
              class="MuiBox-root MuiBox-root-173"
            >
              <h2
                class="MuiTypography-root MuiTypography-subtitle2"
              >
                Notification Name
              </h2>
              <div
                class="MuiBox-root MuiBox-root-174"
              >
                <h6
                  class="MuiTypography-root MuiTypography-subtitle1"
                >
                  Test name
                </h6>
              </div>
              <h2
                class="MuiTypography-root MuiTypography-subtitle2"
              >
                From number
              </h2>
              <div
                class="MuiBox-root MuiBox-root-175"
              >
                <h6
                  class="MuiTypography-root MuiTypography-subtitle1"
                >
                  +34666222333
                </h6>
              </div>
              <h2
                class="MuiTypography-root MuiTypography-subtitle2"
              >
                Notification Message
              </h2>
              <div
                class="MuiBox-root MuiBox-root-176"
              >
                <p
                  class="MuiTypography-root MuiTypography-body2"
                >
                  Test message
                </p>
              </div>
            </dl>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6"
          >
            <h2
              class="MuiTypography-root MuiTypography-subtitle2"
              style="margin-bottom: 15px;"
            >
              Recipients
            </h2>
            <div
              class="MuiBox-root MuiBox-root-179 MuiTableContainer-root makeStyles-tableContainer-177"
            >
              <table
                class="MuiTable-root MuiTable-stickyHeader"
              >
                <thead
                  class="MuiTableHead-root"
                >
                  <tr
                    class="MuiTableRow-root MuiTableRow-head"
                  >
                    <th
                      class="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-head-188 MuiTableCell-sizeSmall MuiTableCell-stickyHeader"
                      scope="col"
                    >
                      Line
                    </th>
                    <th
                      class="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-head-188 MuiTableCell-sizeSmall MuiTableCell-stickyHeader"
                      scope="col"
                    >
                      Recipients Phone Number
                    </th>
                  </tr>
                </thead>
                <tbody
                  class="MuiTableBody-root"
                >
                  <tr
                    class="MuiTableRow-root"
                  >
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
                    >
                      1
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
                    >
                      +34555666777
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div
              class="MuiTablePagination-root"
            >
              <div
                class="MuiToolbar-root MuiToolbar-regular MuiTablePagination-toolbar MuiToolbar-gutters"
              >
                <div
                  class="MuiTablePagination-spacer"
                />
                <p
                  class="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
                >
                  1-1 of 1
                </p>
                <div
                  class="MuiTablePagination-actions"
                >
                  <button
                    aria-label="Previous page"
                    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit Mui-disabled Mui-disabled"
                    disabled=""
                    tabindex="-1"
                    title="Previous page"
                    type="button"
                  >
                    <span
                      class="MuiIconButton-label"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
                        />
                      </svg>
                    </span>
                  </button>
                  <button
                    aria-label="Next page"
                    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit Mui-disabled Mui-disabled"
                    disabled=""
                    tabindex="-1"
                    title="Next page"
                    type="button"
                  >
                    <span
                      class="MuiIconButton-label"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
                        />
                      </svg>
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiPaper-root MuiAlert-root MuiAlert-standardWarning MuiPaper-elevation0"
        role="alert"
      >
        <div
          class="MuiAlert-icon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 5.99L19.53 19H4.47L12 5.99M12 2L1 21h22L12 2zm1 14h-2v2h2v-2zm0-6h-2v4h2v-4z"
            />
          </svg>
        </div>
        <div
          class="MuiAlert-message"
        >
          The time it takes to complete sending all messages depends on the number of recipients and the Twilio account CPS rate (Calls Per Second). Please notice that by default accounts have a 1 CPS rate. Contact Sales or Support to increase your CPS rate.
        </div>
      </div>
    </div>
  </div>
  <div
    class="MuiBox-root MuiBox-root-254"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-8"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Discard
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
        style="text-align: right;"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary"
          style="margin-right: 15px;"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Back
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
          data-testid="send-notification"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Send Notification
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
